### PR TITLE
Update tests for discrete log

### PR DESCRIPTION
**Pull request by sgtlaugh**
## Changes
 - Update tests to include corner cases

### CODE REVIEW

File: pyrival/algebra/discrete_log.py [hunk 1]
```diff
@@ -1,22 +1,50 @@
+from math import gcd, sqrt
+
+
+def extended_gcd(a, b):
+    if not b:
+        return a, 1, 0
+
+    g, y, x = extended_gcd(b, a % b)
+    return g, x, y - ((a // b) * x)
```
Comment (cheran-senthil) [inline pyrival/algebra/discrete_log.py:9]: Please use, https://github.com/cheran-senthil/PyRival/blob/master/pyrival/algebra/gcd.py#L11 for extended_gcd.

File: pyrival/algebra/discrete_log.py [hunk 2]
```diff
@@ -1,22 +1,50 @@
...
-    """returns some x s.t. pow(a, x, m) == b or None if no such x exists, works when a amd m are coprime"""
-    a %= m
-    b %= m
-    if b == 1:
+    """returns smallest non-negative x s.t. pow(a, x, m) == b or None if no such x exists"""
+    if b >= m:
+        return None
+
+    if b == 1 or m == 1:
         return 0
```
Comment (cheran-senthil) [inline pyrival/algebra/discrete_log.py:18]: What if a == 0?

File: pyrival/algebra/discrete_log.py [hunk 3]
```diff
@@ -1,22 +1,50 @@
...
+
+    raw_b, raw_m = b, m
+    c, d, g = 0, 1, gcd(a, m)
+    while g > 1:
+        b, m = b // g, m // g
+        c, d = c + 1, d * (a // g) % m
+        g = gcd(a, m)
+
+    pos = {}
+    p, n = 1, int(sqrt(m - 0.666)) + 1
```
Comment (cheran-senthil) [inline pyrival/algebra/discrete_log.py:34]: What is this `0.666`? Some reference for this would be nice.

File: pyrival/algebra/discrete_log.py (hunk 1 above)
Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py:9]: I'm assuming you mean to import that function and not copy-pasting the exact function used?

I actually considered doing that initially. The reason I did not do so is because in some places like the one below, I've seen existing methods being re-implemented instead of importing. https://github.com/cheran-senthil/PyRival/blob/1971590e96b351288cacb09852896f2383862ab6/pyrival/linear_algebra/multivariable_crt.py#L8 

I'm kinda confused about what pattern we're trying to follow here in this code base. Re-use existing methods whenever possible by importing or keep modules less tightly coupled allowing some duplication. Come to think of it, both has their advantages. Not having code duplication is definitely desirable but at the same time, importing methods from 3 different methods makes it difficult for someone to copy-paste the base method and use it on an actual CP problem.

Will resolve once you clear it up and we can come to a common ground :-)

File: pyrival/algebra/discrete_log.py (hunk 2 above)
Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py:18]: Hmm what about it? Do you have a specific case in mind that is not working as expected? The method should be able to handle when `a=0` and the tests should cover that too ideally.

Checked the implementation and some cases manually, looks all right to me. For example:
`discrete_log(0, 1, 5) = 0`
`discrete_log(0, 0, 5) = 1`

File: pyrival/algebra/discrete_log.py (hunk 3 above)
Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py:34]: Sorry my bad. It's just to avoid precision error. Updating to 0.5 as that will serve the purpose and will make it less alien.

Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py]: Please check now. Should be clearer.

File: pyrival/algebra/discrete_log.py (hunk 1 above)
Comment (cheran-senthil) [reply pyrival/algebra/discrete_log.py:9]: We always reimplement in each file to make copy-pasting during a contest easier. I meant using the code from that file rather than a recursive version which would be slower. Thanks

File: pyrival/algebra/discrete_log.py (hunk 2 above)
Comment (cheran-senthil) [reply pyrival/algebra/discrete_log.py:18]: Hm, this is reasonable. Though taking 0^0 = 1 could be a bit of a stretch on some problems.

File: pyrival/algebra/discrete_log.py [hunk 4]
```diff
@@ -1,22 +1,50 @@
...
+    if b >= m:
+        return None
+
+    if b == 1 or m == 1:
         return 0
 
-    n, vals = int(m**0.5) + 1, {}
-    mult, cur = pow(a, n, m), 1
+    i, t = 1, a
+    while (1 << i) <= m:
```
Comment (cheran-senthil) [inline pyrival/algebra/discrete_log.py:21]: Store `2**i` rather than `i` in a variable and multiply by 2 in each loop.

File: pyrival/algebra/discrete_log.py [hunk 5]
```diff
@@ -1,22 +1,50 @@
...
-
-    cur = b
-    for i in range(n + 1):
-        it = vals.get(cur, None)
-        if it is not None:
-            return it * n - i  # % phi(m)
-        cur *= a
-        cur %= m
+        g, x, y = extended_gcd(d, m)
+        g = ((x * b // g) % m + m) % m
```
Comment (cheran-senthil) [inline pyrival/algebra/discrete_log.py:42]: modulo can't be negative in Python

Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py]: Fixed

File: pyrival/algebra/discrete_log.py (hunk 1 above)
Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py]: Makes sense. Updated.

File: pyrival/algebra/discrete_log.py (hunk 4 above)
Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py:26]: What advantages can we gain storing both? If speed is your concern then firstly this loop will be executed `O(log m)` times at most, where the main complexity of discrete log lies below, i.e `O(sqrt(m))`. And secondly it's not necessarily true storing in another variable will be faster than a simple bit shift with `int`, and even if it were it'd be negligible.

Comment (cheran-senthil) [reply pyrival/algebra/discrete_log.py:26]: Sure, that is reasonable.

File: pyrival/algebra/discrete_log.py [hunk 6]
```diff
@@ -1,22 +1,55 @@
+from math import gcd, sqrt
```
Comment (cheran-senthil) [inline pyrival/algebra/discrete_log.py:1]: Sorry, my review process is slow.

I would prefer `x**0.5` in favor of sqrt and writing gcd out explicitly, like [here](https://github.com/cheran-senthil/PyRival/commit/cfc4643742d5c2fe183d095c7fb2d3a18c8f3d60) is better practice long term for copying code in Python 2.

Comment (bjorn-martinsson) [reply pyrival/algebra/discrete_log.py:1]: From my experience, math.sqrt is both faster and more precise compared to x**0.5. I do however agree with writing out gcd.

Comment (sgtlaugh) [reply pyrival/algebra/discrete_log.py:1]: @cheran-senthil no trouble at all, we have all the time :-)

Don't agree with either of the changes. Not replacing `math.sqrt` with `x**0.5` because of the reason @bjorn-martinsson suggested above and also because it compromises readability. Most importantly, none of these methods are the bottlenecks in this algorithm. No matter how much you optimize `gcd` or `sqrt`, the time saved will only be negligible when compared to the `O(sqrt(m))` dict insertions/lookup that immediately follows. As such I find no good reason to unnecessarily lengthen the implementation.

Note that this was not the case for `extended_gcd` as it was being called `O(sqrt(m))` times. So it made sense to optimize that part.

**Comment by cheran-senthil** [comment]
The tests are still useful, but I think we can stick to @bjorn-martinsson's code for discrete log since there is no gcd overhead there.

**Comment by bjorn-martinsson** [comment]
> The tests are still useful, but I think we can stick to @bjorn-martinsson's code for discrete log since there is no gcd overhead there.

Just a small note about this, when I was reading @sgtlaugh code I thought about how one could make it simpler. After thinking about it I found that there is a super simple way to do it (without ever calling gcd or egcd), you can see it here https://github.com/cheran-senthil/PyRival/blob/master/pyrival/algebra/discrete_log.py . 

Essentially the correctness of my algorithm boils down to this:
Assume p^x is a prime power divisor of mod
then we have 3 cases
1. b divisible by p^x
2. b divisible only by p^y, 0<y<x
3. b is not divisible by p
The important thing is that in case 2, the discrete log (if it exists) cannot be > sqrt(mod),  (technically it should be > log2(mod))
So once you've checked all exponents of a <= sqrt(mod)
You can't have case 2.
Case 2 is the only tricky case.

**Comment by Mukundan314** [comment]
I think splitting the test function into two function named `test_discrete_log` and `test_discrete_log_corner_cases` would be better.

**Comment by sgtlaugh** [comment]
> > The tests are still useful, but I think we can stick to @bjorn-martinsson's code for discrete log since there is no gcd overhead there.
> 
> Just a small note about this, when I was reading @sgtlaugh code I thought about how one could make it simpler. After thinking about it I found that there is a super simple way to do it (without ever calling gcd or egcd), you can see it here https://github.com/cheran-senthil/PyRival/blob/master/pyrival/algebra/discrete_log.py .
> 
> Essentially the correctness of my algorithm boils down to this:
> Assume p^x is a prime power divisor of mod
> then we have 3 cases
> 
> 1. b divisible by p^x
> 2. b divisible only by p^y, 0<y<x
> 3. b is not divisible by p
>    The important thing is that in case 2, the discrete log (if it exists) cannot be > sqrt(mod),  (technically it should be > log2(mod))
>    So once you've checked all exponents of a <= sqrt(mod)
>    You can't have case 2.
>    Case 2 is the only tricky case.

Just took a look. That should work and it's definitely better and simpler.

**Comment by sgtlaugh** [comment]
> The tests are still useful, but I think we can stick to @bjorn-martinsson's code for discrete log since there is no gcd overhead there.

Sure, updated the PR
